### PR TITLE
CDAP-7930 Fix incorrect metadata search total

### PR DESCRIPTION
**Pull request by Denton-L**
Fixes the following:

* An incorrect total would be returned if an offset > number of results
  was specified while doing a query with non-default sort order.

* If offset > number of results, limit does not restrict the number of
  results appropriately.

* ArrayList was being used where LinkedList would be more efficient.

* A negative offset or limit was unchecked and could potentially cause
  problems.

JIRA: https://issues.cask.co/browse/CDAP-7930
Bamboo: http://builds.cask.co/browse/CDAP-DUT5420-1

### CODE REVIEW

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/store/DefaultMetadataStore.java [hunk 1]
```diff
@@ -442,9 +440,10 @@ private MetadataSearchResponse search(Set<MetadataScope> scopes, String namespac
     Map<NamespacedEntityId, Metadata> systemMetadata = fetchMetadata(sortedEntities, MetadataScope.SYSTEM);
     Map<NamespacedEntityId, Metadata> userMetadata = fetchMetadata(sortedEntities, MetadataScope.USER);
 
+    Set<MetadataSearchResultRecord> resultRecords = addMetadataToEntities(sortedEntities, systemMetadata, userMetadata);
     return new MetadataSearchResponse(
-      sortInfo.getSortBy() + " " + sortInfo.getSortOrder(), offset, limit, numCursors, total,
-      addMetadataToEntities(sortedEntities, systemMetadata, userMetadata), cursors, showHidden
+      sortInfo.getSortBy() + " " + sortInfo.getSortOrder(), offset, limit, numCursors, resultRecords.size(),
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/store/DefaultMetadataStore.java]: Is this correct? If in total there are 100 records, if request offset=0 and limit=10, then `resultRecords.size()` is only `10` instead of `100`, right? I thought total means total number of records, i.e. something won't change regardless of what `limit` and `offset` is.

Comment (Denton-L) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/store/DefaultMetadataStore.java]: Originally, it `total` would be 100. In this change, the number will now be `10`. I'm not sure if this is the "correct" behaviour but this is what the JIRA requested.

If this is incorrect, I can close the JIRA as won't fix.

Comment (chtyim) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/store/DefaultMetadataStore.java]: @bdmogal can you clarify the requirement? The JIRA is not very clear to me.

Comment (bdmogal) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/store/DefaultMetadataStore.java]: This seems wrong. Search should **always** return the total number of entities. @anwar6953 suspected that there is a bug in the total, when you specify *offset* > *number of results*. That's what needs to be verified, I think.

Like @chtyim said, when there are 100 records, if offset = 0 and limit = 10, total should remain 100, because that helps the UI calculate the number of pages.

Comment (Denton-L) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/store/DefaultMetadataStore.java]: I messaged him on Slack before starting this ticket and he said this:

>if giving an offset of 10, that means it should skip the first 10 entries
>and also not include those in the ‘total’ number

So I guess this ticket should be closed as won't fix because we are, indeed, returning the correct total.

Comment (anwar6953) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/store/DefaultMetadataStore.java]: Yea, I suspect an issue to be there when the offset is a large number (larger than the number of actual results).

Comment (anwar6953) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/store/DefaultMetadataStore.java]: @Denton-L can you verify what the 'total' is, when specifying an offset that is much larger than the actual total?

Comment (Denton-L) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/store/DefaultMetadataStore.java]: It seems to return the same total, regardless of what offset you specify.

Comment (Denton-L) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/store/DefaultMetadataStore.java]: Actually, revisiting it, I think there actually is a bug for when you provide a `sort` parameter... I'll look into it.

Comment (anwar6953) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/store/DefaultMetadataStore.java]: That's with this fix, right?
And without the fix, the total is much higher (depending on the offset provided)?

Comment (Denton-L) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/store/DefaultMetadataStore.java]: Currently, without the fix, the query will return a correct total for any offset given that the `sort` param is not given. If the `sort` param is given, the total is incorrect without the fix.

I am currently writing a new patch to fix this because I think I've found the root of the problem.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java [hunk 2]
```diff
@@ -468,7 +468,7 @@ private Metadata getSnapshotBeforeTime(NamespacedEntityId targetId, long timeMil
       return Collections.emptySet();
     }
 
-    List<ImmutablePair<byte [], byte []>> fuzzyKeys = new ArrayList<>();
+    List<ImmutablePair<byte [], byte []>> fuzzyKeys = new LinkedList<>();
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java]: Why the change? Also, since we know the expected size (`targetIds.size()`), we should use `new ArrayList<>(targetIds.size())`.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java [hunk 3]
```diff
@@ -563,15 +563,14 @@ private MetadataEntry convertRow(Row row) {
...
-    return searchByCustomIndex(namespaceId, types, sortInfo, offset, limit, numCursors, cursor, showHidden);
+    return SortInfo.DEFAULT.equals(sortInfo) ?
+      searchByDefaultIndex(namespaceId, searchQuery, types, showHidden) :
+      searchByCustomIndex(namespaceId, types, sortInfo, offset, limit, numCursors, cursor, showHidden);
   }
 
   private SearchResults searchByDefaultIndex(String namespaceId, String searchQuery,
                                              Set<EntityTypeSimpleName> types, boolean showHidden) {
-    List<MetadataEntry> results = new ArrayList<>();
+    List<MetadataEntry> results = new LinkedList<>();
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java:573]: Why the change?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/store/DefaultMetadataStore.java [hunk 4]
```diff
@@ -398,44 +399,44 @@ private MetadataSearchResponse search(Set<MetadataScope> scopes, String namespac
...
+    List<MetadataEntry> results = new LinkedList<>();
+    List<String> cursors = new LinkedList<>();
+    List<MetadataEntry> allResults = new LinkedList<>();
     for (MetadataScope scope : scopes) {
       SearchResults searchResults =
         getSearchResults(scope, namespaceId, searchQuery, types, sortInfo, offset, limit, numCursors, cursor,
                          showHidden);
       results.addAll(searchResults.getResults());
       cursors.addAll(searchResults.getCursors());
+      allResults.addAll(searchResults.getAllResults());
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/store/DefaultMetadataStore.java:419]: Why storing all results??

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java (hunk 3 above)
Comment (Denton-L) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java:573]: Since we're not doing any random-access and we don't know the size of the results prior to creation, I feel like it would be more efficient to use a `LinkedList` over an `ArrayList`. Please let me know if I'm wrong about this.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/store/DefaultMetadataStore.java (hunk 4 above)
Comment (Denton-L) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/store/DefaultMetadataStore.java:419]: We need to get all of the results so that we can count them in the total, I believe.

**Comment by Denton-L** [comment]
@chtyim I've tried to address all of the comments.

**Comment by sreevatsanraman** [comment]
LGTM
